packages/libraries/sdk-rs/src/supergraph_fetcher.rs

```diff
@@ -1,86 +1,241 @@
+use std::fmt::Display;
+use std::sync::RwLock;
 use std::time::Duration;
 
+use reqwest::header::HeaderMap;
+use reqwest::header::HeaderValue;
+use reqwest::header::InvalidHeaderValue;
+use reqwest::header::IF_NONE_MATCH;
+use reqwest_middleware::ClientBuilder;
+use reqwest_middleware::ClientWithMiddleware;
+use reqwest_retry::policies::ExponentialBackoff;
+use reqwest_retry::RetryTransientMiddleware;
+
+#[derive(Debug)]
+pub struct SupergraphFetcherAsyncClient {
+    reqwest_client: ClientWithMiddleware,
+}
+
 #[derive(Debug)]
-pub struct SupergraphFetcher {
-    client: reqwest::blocking::Client,
+pub struct SupergraphFetcherSyncClient {
+    reqwest_client: reqwest::blocking::Client,
+    retry_count: u32,
+}
+
+#[derive(Debug)]
+pub struct SupergraphFetcher<AsyncOrSyncClient> {
+    client: AsyncOrSyncClient,
     endpoint: String,
-    key: String,
-    user_agent: String,
-    etag: Option<String>,
+    etag: RwLock<Option<HeaderValue>>,
+}
+
+pub enum SupergraphFetcherError {
+    FetcherCreationError(reqwest::Error),
+    NetworkError(reqwest_middleware::Error),
+    NetworkResponseError(reqwest::Error),
+    HeadersLock(String),
+    InvalidKey(InvalidHeaderValue),
+}
+
+impl Display for SupergraphFetcherError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            SupergraphFetcherError::FetcherCreationError(e) => {
+                write!(f, "Creating fetcher failed: {}", e)
+            }
+            SupergraphFetcherError::NetworkError(e) => write!(f, "Network error: {}", e),
+            SupergraphFetcherError::NetworkResponseError(e) => {
+                write!(f, "Network response error: {}", e)
+            }
+            SupergraphFetcherError::HeadersLock(e) => write!(f, "Headers lock error: {}", e),
+            SupergraphFetcherError::InvalidKey(e) => write!(f, "Invalid CDN key: {}", e),
+        }
+    }
+}
+
+fn prepare_endpoint_and_header(
+    mut endpoint: String,
+    key: &str,
+) -> Result<(String, HeaderMap), SupergraphFetcherError> {
+    if !endpoint.ends_with("/supergraph") {
+        if endpoint.ends_with("/") {
+            endpoint.push_str("supergraph");
+        } else {
+            endpoint.push_str("/supergraph");
+        }
+    }
+
+    let mut headers = HeaderMap::new();
+    headers.insert(
+        "X-Hive-CDN-Key",
+        HeaderValue::from_str(key).map_err(SupergraphFetcherError::InvalidKey)?,
+    );
+
+    Ok((endpoint, headers))
 }
 
-impl SupergraphFetcher {
+impl SupergraphFetcher<SupergraphFetcherSyncClient> {
```

Comment:
`try_new` can move into `impl SupergraphFetcher {` so it will be:

```
SupergraphFetcher::new_async()
SupergraphFetcher::new_sync() 
```

And then the consumer doesn't need to know about the type-state.